cpp/src/io/shp/polygon_shapefile_reader.cu

```diff
@@ -0,0 +1,244 @@
+    {
+        std::vector<int> g_len_v,f_len_v,r_len_v;
+        std::vector<double> x_v, y_v;
+        GDALAllRegister();
+        
+        GDALDatasetH hDS = GDALOpenEx( filename, GDAL_OF_VECTOR, NULL, NULL, NULL );
+        CUDF_EXPECTS(hDS!=NULL,"Failed to open ESRI Shapefile dataset");		    
+        OGRLayerH hLayer = GDALDatasetGetLayer( hDS,0 );
+        CUDF_EXPECTS(hLayer!=NULL,"Failed to open the first layer");
+        int num_f=ReadLayer(hLayer,g_len_v,f_len_v,r_len_v,x_v,y_v);
+        CUDF_EXPECTS(num_f>0,"Shapefile must have at lest one polygon");
+        
+        pm.num_group=g_len_v.size();
+        pm.num_feature=f_len_v.size();
+        pm.num_ring=r_len_v.size();
+        pm.num_vertex=x_v.size();
+        pm.group_length=new uint32_t[ pm.num_group];
+        pm.feature_length=new uint32_t[ pm.num_feature];
+        pm.ring_length=new uint32_t[ pm.num_ring];
+        pm.x=new double [pm.num_vertex];
+        pm.y=new double [pm.num_vertex];
+        CUDF_EXPECTS(pm.group_length!=nullptr&&pm.feature_length!=nullptr&&pm.ring_length!=nullptr,"expecting p_{g,f,r}_len are non-zeron");		    
+        CUDF_EXPECTS(pm.x!=nullptr&&pm.y!=nullptr,"expecting polygon x/y arrays are not nullptr");
+	      
+        memcpy((void *)(pm.group_length),(void *)(g_len_v.data()),pm.num_group*sizeof(uint32_t));
+        memcpy((void *)(pm.feature_length),(void *)(f_len_v.data()),pm.num_feature*sizeof(uint32_t));
+        memcpy((void *)(pm.ring_length),(void *)(r_len_v.data()),pm.num_ring*sizeof(uint32_t));
+        memcpy((void *)(pm.x),(void *)(x_v.data()),pm.num_vertex*sizeof(double));
+        memcpy((void *)(pm.y),(void *)(y_v.data()),pm.num_vertex*sizeof(double));      
+    }
+
+    /*
+    * read polygon data from file in ESRI Shapefile format; data type of vertices is fixed to double (GDF_FLOAT64)
+    * see shp_readers.hpp
+    */
+
+    void read_polygon_shapefile(const char *filename,
+                      gdf_column* ply_fpos, gdf_column* ply_rpos,
+                      gdf_column* ply_x, gdf_column* ply_y)
+    {
+        memset(ply_fpos,0,sizeof(gdf_column));
+        memset(ply_rpos,0,sizeof(gdf_column));
+        memset(ply_x,0,sizeof(gdf_column));
+        memset(ply_y,0,sizeof(gdf_column));
+
+        struct polygons<double> pm;
+        memset(&pm,0,sizeof(pm));
+        polygon_from_shapefile(filename,pm);
+        if (pm.num_feature <=0) return;
+
+        cudaStream_t stream{0};
+        auto exec_policy = rmm::exec_policy(stream)->on(stream);
+
+        int32_t* temp{nullptr};
+        RMM_TRY( RMM_ALLOC(&temp, pm.num_feature * sizeof(int32_t), stream) );
+        CUDA_TRY( cudaMemcpyAsync(temp, pm.feature_length,
+                              pm.num_feature * sizeof(int32_t),
+                              cudaMemcpyHostToDevice, stream) );
+        //prefix-sum: len to pos
+        thrust::inclusive_scan(exec_policy, temp, temp + pm.num_feature, temp);
```

Comment:
```suggestion
        thrust::inclusive_scan(exec_policy, pm.feature_length, pm.feature_length + pm.num_feature, temp);
```

And delete the preceding cudaMemcpyAsync.

Reply:
pm.feature_length is a host pointer, not a device pointer. Is Thurst supporting scan (and maybe other ops) host vectors into device vectors directly?